core/src/main/java/com/sequenceiq/cloudbreak/service/stack/DefaultStackService.java

```diff
@@ -195,6 +196,28 @@ public void delete(Long id, CbUser user) {
     }
 
     @Override
+    public void removeInstance(CbUser user, Long stackId, String instanceId) {
+        Stack stack = stackRepository.findById(stackId);
+        InstanceMetaData instanceMetaData = instanceMetaDataRepository.findByInstanceId(stackId, instanceId);
+        if (stack == null) {
+            throw new NotFoundException(String.format("Not found stack with id '%s'", stackId.toString()));
+        }
+        if (instanceMetaData == null) {
+            throw new NotFoundException(String.format("Not found instance metadata with id '%s'", instanceId));
+        }
+        if (stack.getCluster() == null) {
+            throw new NotFoundException(String.format("Not found cluster for stack '%s'", stack.getId()));
+        }
+        if (!stack.getAccount().equals(user.getAccount())) {
+            throw new BadRequestException(String.format("Stack %s is not accessible from this account.", stackId));
+        }
+        if (!stack.isPublicInAccount() && !stack.getOwner().equals(user.getUserId())) {
+            throw new BadRequestException(String.format("Private stack (%s) only modifiable by the owner.", stackId));
+        }
```

Comment:
If the findOne method is used instead of findById, it is not needed to check the user and the owner because the OwnerBasedPermissionEvaluator class does this through the PostAuthorize annotation.